Rewrite timers to be more specification compliant

mstoykov opened the pull request:
This is more or less a complete rewrite following the specification including leaving links to relevant parts and comments on why we diverge.

Among other changes are:
1. timeout/intervals that triggered but weren't executed before they were cleared are *not* executed.
2. setInterval gets rescheduled after it's execution as per the specification.
3. setTimeout/setInterval triggers are ordered based on their invocation as well as the timeout. Again as per the specification.

Fixes #3

mstoykov committed:
Rewrite timers to be more specification compliant

This is more or less a complete rewrite following the specification
including leaving links to relevant parts and comments on why we
diverge.

Among other changes are:
1. timeout/intervals that triggered but weren't executed before they
   were cleared are *not* executed.
2. setInterval gets rescheduled after it's execution as per the
   specification.
3. setTimeout/setInterval triggers are ordered based on their invocation
   as well as the timeout. Again as per the specification.

Fixes #3

olegbespalov commented on timers/timers.go:
```diff
@@ -20,8 +18,16 @@ type Timers struct {
+	// it is just a list of the id in their time.Time order.
+	// it is used to get timers fire in sequence.
+	// not anything more then a slice as it is unlikely it will have too many ids to begin with.
+	timersQueue []int
```
Out of curiosity, why not `uint32` ? Using this type, we can significantly reduce the number of places where we currently do a conversion, no? :thinking:

mstoykov replied:
I was mostly thinking of moving us away from uint32 ... but didn't end up doing it.

I am not certain why we are saving 32bits just to turn out someone ends up using 4 billion timers over the course of one test run.

So I will probably take another pass and bump it to 64bit :thinking:

olegbespalov replied:
Sure, if it will be exact 64 bits everywhere, it's also fine :+1:

mstoykov replied:
fixed as part of [a91e635](https://github.com/grafana/xk6-timers/pull/7/commits/a91e6354a3f9944db8d243c29409e788f22d84aa)

olegbespalov commented on timers/timers.go:
```diff
@@ -85,70 +72,146 @@ func (e *Timers) call(callback goja.Callable, args []goja.Value) error {
+
+	task := func() error {
+		if _, exist := e.timers[id]; !exist {
+			return nil // 8.1
```
```suggestion
                // Specification 8.1: If id does not exist in global's map of active timers, then abort these steps.
		if _, exist := e.timers[id]; !exist 
			return nil
```

olegbespalov reviewed (commented):
It mostly looks good :+1: I left minor comments. 

Out of curiosity, since the internals of the timers module were rewritten, are you still going to make xk6 timers part of the core in the following k6 release?

mstoykov committed:
Update timers/timers.go

Co-authored-by: Oleg Bespalov <oleg.bespalov@grafana.com>

mstoykov commented:
> Out of curiosity, since the internals of the timers module were rewritten, are you still going to make xk6 timers part of the core in the following k6 release?

Arguably the experimental modules are for:
1. figuring out the API - we do not have to this here at all
2. figuring out if we will add it - I am pretty sure here we have no problem as well.
3. figuring out behaviour - we do change this here, but we do it towards what is actually expected.


So arguably this API never needed to be an experimental one. But due to history we did start it as an extension, and then when exprimental modules started existing we just added it as it is pretty good candidate.

There are still some things to think of such as whether they should be in a module at all.

So - yes, I still think we can add it in whichever release we decide. The underlying implementation may or may not have bugs, but we will still need to fix those.
For now - lets finish it and hten lets see if we have any problems.

mstoykov committed:
fixup! Update timers/timers.go

oleiade reviewed (approved):
Great work 👌🏻 , I really have nothing to bring up 🚀 🚀

mstoykov committed:
Rewrite the implementation again

1. split slice manipulation code to a struct
2. more types for less bookkeeping
3. use the taskqueue for less likely races
4. use uint64 for all timer ids instead of uint32

mstoykov commented:
Sorry about this, but I started pulling out some stuff in types (mostly to not keep manipulating slices everywhere) and it did take me longer then expected to split the fields around.

And I did introduce a bug that took me a while to run down. 

The new code IMO is way more readable as all the slice stuff is moved away and uses the taskqueue as to be even more sure I do not have races.

mstoykov committed:
fixup! Rewrite the implementation again

olegbespalov commented on timers/timers.go:
```diff
@@ -55,100 +62,211 @@ func (e *Timers) Exports() modules.Exports {
-	id, stopCh := e.getTimerStopCh()
+func (e *Timers) setupTaskTimeout() {
+	e.queue.stopTimer()
+	delay := -time.Since(e.timers[e.queue.first().id])
```
Not sure that it's easy to get what is happening here, maybe let's drop a comment? and the follow-up will be first() always non-empty, I mean could we ended up with `0` or it simply fine since in that case wan't be produced any delay and the task is free to go?

mstoykov replied:
In practice everywhere where this happens we either:
1. check the queue is not empty
2. we just added a timer to the queue
